### DISCUSSION TITLE
Interface improvements

### DISCUSSION THREAD

**Discussion by ValerianRey** [TorchJD/torchjd#507] [General]
I think the current interface of torchjd has a few issues:
- The role of `backward` is clear in torch: it computes and populates the `grad` fields. In torchjd, we have 2 possibilities:
  - 1] Populate the `grad` fields with the aggregation of the Jacobian. This is what we currently do. In the Gramian-based approach, that would be with the gradient of the weighted sum of losses.
  - 2] Depending on whether we use `autojac` or `autogram`, we would populate some `jac` fields with the Jacobian itself or some `gram` fields with the (non-summed) gramians.
  
  So the question is: do we want `backward` to "feed" the optimizer (always populate directly the `grad` field), or do we want `backward` to populate whatever is at its abstraction level? We currently go for option 1 but I have a few arguments in favor of option 2:
  - `autojac.backward` and `autojac.mtl_backward` both take an aggregator. I think this is kind of mixing the responsibility of the efficient computation of the Jacobian (`torchjd.autojac`) with that of the aggregation (`torchjd.aggregation`). In an ideal world, I would prefer the `autojac` package to be entirely independent of the `aggregation` package.
  - It would be much easier to understand what's going on from a user's perspective, if we split the computation of the Jacobian and its aggregation into the equivalent of a gradient, into two different steps.
  - It would make it possible to monitor some stats about the Jacobian without having to use hooks.
  - It would be possible to customize what someone does with those `jac` fields. For instance, someone may want to use two different aggregators for two different subsets of the parameters. It's currently not possible at all because `backward` takes a single `Aggregator`.

  I also see three arguments against option 2:
  - It would add the need of an object responsible for transforming the `jac` fields into `grad` fields. We can call it `JacAggregator` in this discussion.
  - We would need a `JacAggregator.zero_jac()` method to free the `jac` fields similarly as `optimizer.zero_grad()` frees the `grad` fields.
  - It makes JD doable in 3 lines of code instead of one: `autojac.backward(...)  # compute and populate the jac fields`, `jac_aggregator.step()  # aggregate the jacobian matrix from the jac fields to populate the grad fields` and `jac_aggregator.zero_jac()  # free the jac fields`.

- `autojac.backward` is the analog to `autograd.backward`, but we don't have an analog to `autograd.grad`. This is being fixed in #505 with `autojac.jac`.
- There is a discussion about whether `autojac.backward`, `autojac.mtl_backward` and `autojac.jac` should take `grad_outputs` or `jac_outputs` arguments. I think this is highly tied to the above discussion (option 1 vs option 2). With option 1, since we fill `grad` fields at the end, it would make sense to also take grads (`grad_outputs`) at the beginning. With option 2, since we stay at our own abstraction level and fill `jac` fields at the end, it would make sense to also take jacs (`jac_outputs`) at the beginning). Once again, staying at our own abstraction level makes us more expressive (because even non-diagonal `jac_outputs` would be allowed) but would require an extra line of code for some users (to diagonalize some `grad_outputs` into some `jac_outputs`).
- `torchjd.autogram` should have a similar interface to that of `torchjd.autojac`, i.e. a `backward` and a `gram` function. This would be easy to fix with #476, because if we stop relying on the `batch_dim` param, we stop needing the `Engine` altogether.

**Comment by PierreQuinton** [comment]
I think the main thing here is the dependence of `autojac` on `aggregation`: This hints bad design, I really like option 2. I think it also allows for framing mtl_backward differently as we would then need a function that computes task specific grads, and accumulate activations `jac` field, we can then use backward normally with this jacobian. This is much more atomic, and then mtl_backward is less of the monstruosity it currently is.

**Reply by ValerianRey** [reply]
> and accumulate activations `jac` field

The thing is, once it's accumulated into them, it's not very practical to use. We would need to collect them into a tuple and free the fields:
```python
# Have a function filling grad field of task-specific and jac field of shared
# ...
jac_outputs = tuple(f.jac for f in features)
for f in features:
    f.jac = None
# ...
# Use jac_outputs in a last call to autojac.backward
```

Alternatively, we could either have this function return both grads and jacobians, or have this function fill into grad and return jacobians wrt activations, but I think it's quite complex.

So I think I would rather keep mtl_backward similar to what it is. Of course, it would still be much more simple with option 2 than with option 1, because it would not be responsible for the aggregation. So it would just:
- Compute gradients wrt task-specific params and populate the grad fields with them
- Compute the jacobians wrt to shared params and populate the jac fields with them

The code would be:
```python
# Earlier, outside of the training loop
aggregator = UPGrad()
jac_aggregator = JacAggregator(shared_model.parameters(), aggregator)

mtl_backward(losses, features)  # fill grad field of task-specific params, fill jac field of shared params
jac_aggregator.step()  # fill grad fields of shared params using jac fields
jac_aggregator.zero_jac()  # free jac fields of shared params
```

**Comment by ValerianRey** [comment]
I'm not sure we actually need a `JacAggregator` object. I think a simple function works too. I would maybe call it `jac_to_grad`.
So the standard JD step would look like:
```python
criterion = MSELoss(reduction="none")
model = MyModel()
optimizer = SGD(model.parameters())
aggregator = UPGrad()

for input, target in dataloader:
    output = model(input)
    losses = criterion(output, target)

    autojac.backward(losses)  # fill jac fields
    jac_to_grad(model.parameters(), aggregator)  # fill grad fields
    zero_jac(model.parameters())
    optimizer.step()
    optimizer.zero_grad()
```

And the gramian-based JD step would look like:
```python
criterion = MSELoss(reduction="none")
model = MyModel()
optimizer = SGD(model.parameters())
weighting = UPGradWeighting()

for input, target in dataloader:
    output = model(input)
    losses = criterion(output, target)

    gramian = autogram.gram(losses)
    weights = weighting(gramian)
    losses.backward(weights)
    optimizer.step()
    optimizer.zero_grad()
```

If someone somehow wants to compute the Jacobian without actually doing JD:
```python
criterion = MSELoss(reduction="none")
model = MyModel()
optimizer = SGD(model.parameters())
aggregator = UPGrad()

for input, target in dataloader:
    output = model(input)
    losses = criterion(output, target)

    jacobians = autojac.jac(losses)
    jacobian_matrix = torch.concat([jacobian.reshape(jacobian.shape[0], -1)], dim=1)  # this could be a utility function
    # do whatever with this
```

Lastly, if someone wants to have the individual gramians in the `gram` fields (note that this could be useful to get an idea of how each parameter contributes to the conflict, or to the gradient norm):
```python
criterion = MSELoss(reduction="none")
model = MyModel()
optimizer = SGD(model.parameters())
weighting = UPGradWeighting()

for input, target in dataloader:
    output = model(input)
    losses = criterion(output, target)

    autogram.backward(losses)  # fills the .gram fields
    # maybe do zero_gram(model.parameters())?
```

I also have a feeling that `autogram.gram` should compute one gramian per param like `autojac.jac` computes on jacobian per param and `autograd.grad` computes one grad per param. I don't think we save that much memory by summing the gramians as soons as we can. In fact, even in torch, they could technically use the gradients wrt individual parameter tensors as soon as they are computed, without waiting for the full backward pass to be completed, but they don't because it would be an architectural nightmare. My point is that even when it does actually take quite a lot of memory, torch doesn't do this optimization that we currently do. Assuming float32 gramians of shape 64x64 with 1000 parameter tensors, we save ~16MB of memory by doing that. I really don't think it's worth it. So IMO we should have `autogram.gram` return a tuple of gramians.

Then, the standard gramian-based JD step would look like:
```python
criterion = MSELoss(reduction="none")
model = MyModel()
optimizer = SGD(model.parameters())
weighting = UPGradWeighting()

for input, target in dataloader:
    output = model(input)
    losses = criterion(output, target)

    gramians = autogram.gram(losses)
    gramian = torch.stack(gramians, dim=0).sum(dim=0)
    weights = weighting(gramian)
    losses.backward(weights)
    optimizer.step()
    optimizer.zero_grad()
```

**Comment by ValerianRey** [comment]
Here are the first few steps I have in mind:
- Add a `jac_to_grad` function in `autojac`
- Add a `zero_jac` function in `autojac`
- Change the `Accumulate` transform to accumulate jacobians into jac fields rather than gradients into grad fields
- Remove the `Aggregate` transform (some of its code will be needed in `jac_to_grad` so it should be moved there)
- Change `mtl_backward` and `backward` to not take an aggregator
- Change usage examples and see if it looks good in practice

**Reply by PierreQuinton** [reply]
Very nice. I think we should dismantle poco a poco the Transform package. We will end up dropping most of it in favor of functions.

**Comment by ValerianRey** [comment]
I think I'll get rid of the `zero_jac` function, and add a boolean parameter `zero_jac: bool = True` to `jac_to_grad`.
So we'll be able to do:
```python
autojac.backward(losses)
jac_to_grad(model.parameters(), aggregator)
```

instead of 
```python
autojac.backward(losses)
jac_to_grad(model.parameters(), aggregator)
zero_jac(model.parameters())
```

and we can still choose to keep the jacs by doing:
```python
autojac.backward(losses)
jac_to_grad(model.parameters(), aggregator, zero_jacs=False)
```

In those quite rate cases, it's still quite easy for the user to free the jacs themselves, even if we don't provide a utility function for this:
```python
for param in model.parameters():
    param.jac = None
```

Note that we could use a different name than `zero_jacs`, like `retain_jacs` or `free_jacs`. But in any case I'd like to avoid having an extra function to call and an extra import to make.

**Comment by PierreQuinton** [comment]
I think the reason for zero_jac would be that jacobians are accumulated, otherwise there is no problem.

**Comment by ValerianRey** [comment]
@PierreQuinton I think you raised an important question in this PR comment, so I'd like to continue the discussion here.

https://github.com/TorchJD/torchjd/pull/525#issuecomment-3854396646

> Following our discussion where we considered making the Generics `Weighting[PSDMatrix]` and `Weighting[Matrix]` into abstract classes which we can then use to differentiate the method used for `jac_to_grad`, we could in principle avoid testing instances and rather have methods in each specifying how we handle a list of Jacobians. This is good in terms of not have cases, but is bad in terms of having a less clear cut algebraic meaning to the aggregation package.

I think that if we go in this direction, we could have very flexible aggregators, that would take jacobians (not necessarily matrices) as inputs. They would also be able to take an iterable of jacobians.

So the prototype would go from:
```python
def __call__(matrix: Matrix) -> Vector
```
(i'm using types that don't explicitly exist or that are not public but this is the reality of what we're doing)

To:

```python
def __call__(tensors: Tensor | Iterable[Tensor]) -> Tensor | Iterable[Tensor]
```
With the necessary overloads, we would be able to indicate that if a single tensor is provided, a single tensor is output. And if an iterable of tensors is provided, an iterable of tensors is output.

In that case, the aggregator would be responsible for most of what the jac_to_grad method currently does. Basically, jac_to_grad would just store into .grad the aggregated .jac fields, and free the .jac.

I think the aggregator would still be able to do everything that it was responsible for in the past, but it will now also have more uses. All of this still with a single `__call__` method. I don't think I'm in favor of having a different function to handle iterables of jacobians, or to handle non-matrix jacobians. This would be quite confusing.

I'm not sure it's a good idea, but it's worth considering.

An argument in favor of that, is that in maths, a jacobian is of shape Rmxn, and a gradient is of shape Rn, but in torch, a gradient is of shape Rn1x...xnk (sorry for the terrible formatting, I'm not even trying). So why would we need to translate from torch to math (flatten into jacobian matrix) to aggregate, and then translate back from math to torch (reshape into parameter shape)? I think it doesn't make a lot of sense. Especially since jacobian matrices are jacobians. So the proposed interface of aggregators will still be able to aggregate jacobians matrices. We sometimes do need to have matrices in order to apply the aggregator's algorithm. But for instance, it's not the case for gramian weighted aggregators : we can directly compute the matrix gramian from a non-matrix jacobian. So there's no point in explicitly reshaping the jacobia  into matrix. And for aggregators in which it's necessary to work on a matrix, we can just reshape precisely when we need it, and not before.

**Reply by PierreQuinton** [reply]
I'm liking this direction. I think we need to be slightly critical still. A very good thing is that this is backward compatible on the aggregation package. Also this release will start looking like something truly major.
